Added a new suite of tools for variant filtering based on site-level annotations.

samuelklee opened the pull request:
This adds the following tools, which supplant the VQSR workflow: ExtractVariantAnnotations, TrainVariantAnnotationsModel, and ScoreVariantAnnotations. See meta issue https://github.com/broadinstitute/gatk/issues/7724.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java:
```diff
@@ -34,6 +34,7 @@ public final class MathUtils {
     public static final double LOG10_ONE_HALF = Math.log10(0.5);
     public static final double LOG10_ONE_THIRD = -Math.log10(3.0);
     public static final double LOG_ONE_THIRD = -Math.log(3.0);
+    public static final double LOG_2 = Math.log(2.0);
```
This and the below came from the BGMM branch, I'll leave them in for now.

samuelklee commented:
I still need to finish up the tool-level Javadocs for the TrainVariantAnnotationsModel tool. But since I'll be off on vacation until the end of the week, I wanted to go ahead and open this up for review.

There's a lot here, but not too much of it is production code (<2k LOC). I've split things up into commits that should hopefully make it more easy to review. The first commit contains the WDL added in #7932 and has already been reviewed by me, although it may benefit from a second pass. The second commit updates that WDL to account for some changes I added after review.

There are TODOs scattered throughout the code, but some of them are intentionally left as an exercise for future developers. See the meta issue linked above to get an idea of what might be appropriate to leave to future work.

There are also stubs throughout for the BGMM implementation, which will be added in a separate PR. Hopefully we can get some ML club reviewers then.

@meganshand @droazen @davidbenjamin mind taking a look or suggesting other reviewers? I would hope that we can get this in by the next release after the other flow-based methods are released, since the IsolationForest filtering method added here is also used in that pipeline. It would also be nice to get this merged by the next release to keep us on track on the malaria side.

samuelklee commented on build.gradle:
```diff
@@ -285,6 +285,7 @@ dependencies {
 
     implementation 'org.apache.commons:commons-lang3:3.5'
     implementation 'org.apache.commons:commons-math3:3.5'
+    implementation 'org.hipparchus:hipparchus-stat:2.0'
```
See https://github.com/broadinstitute/dsp-methods-model-prototyping/discussions/9#discussioncomment-2290858.

samuelklee commented:
Also, if the WDL-generation and tab-completion tests continue to fail, I’ll address it after I get back. I think this has something to do with non-ASCII characters in the Javadocs, but I thought I had gotten all of them…

meganshand commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotationsIntegrationTest.java:
```diff
@@ -0,0 +1,253 @@
+                EXPECTED_TEST_FILES_DIR,
+                tag + LabeledVariantAnnotationsWalker.ANNOTATIONS_HDF5_SUFFIX,
+                outputPrefix + LabeledVariantAnnotationsWalker.ANNOTATIONS_HDF5_SUFFIX));
+        SystemCommandUtilsTest.runSystemCommand(String.format("diff %s/%s.vcf %s.vcf",
```
I'm surprised you don't need to strip out the header from these files before doing the diff. Does that mean that none of these tools add a program line to their output VCFs? They probably should if they aren't currently.

samuelklee replied:
Oops, good catch. I'll use `getDefaultToolVCFHeaderLines` but then set `addOutputVCFCommandLine` to false for all tests, so long as we're OK with not testing for the presence of these lines.

samuelklee replied:
Haha actually wait, this is already how it is!

meganshand replied:
Oops sorry, totally missed that.

meganshand reviewed (commented):
This looks great @samuelklee! I thought the layout of classes was very clear and the testing was thorough. I had a few questions and comments. Also, I noticed a few places where there were temp files to read and write data within the tool and I was surprised that was necessary.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/LabeledVariantAnnotationsWalker.java:
```diff
@@ -0,0 +1,382 @@
+        return null;
+    }
+
+    // TODO maybe clean up all this Triple and metadata business with a class?
```
Is there a compelling need to defer this TODO?  It seems worth doing now, IMO.

samuelklee replied:
Hmm, after thinking about it more, I removed the TODO and left it as is. This sort of need for a single-use container seems like the perfect time to use an N-tuple.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java:
```diff
@@ -0,0 +1,625 @@
+ *     This tool is intended to be used as the last step in a variant-filtering workflow that supersedes the
+ *     {@link VariantRecalibrator} workflow. Using a previously trained model produced by {@link TrainVariantAnnotationsModel},
+ *     this tool assigns a score to each call (with a lower score indicating that a call is more likely to be an artifact).
+ *     Each score can also be converted to a corresponding sensitivity to a calibration set, if the latter is available.
```
"sensitivity to" --> "sensitivity with respect to"

samuelklee replied:
Thanks, changed everywhere (I think).

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/TrainVariantAnnotationsModel.java:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    private static int numPassingFilter(List<Boolean> isPassing) {
+        return isPassing.stream().mapToInt(x -> x ? 1 : 0).sum();
```
Use `Stream::count`

samuelklee replied:
OK, changed to `(int) isPassing.stream().filter(x -> x).count()` (and made `isPassing` final).

davidbenjamin reviewed (approved):
This looks really clean.  The documentation is excellent.

samuelklee commented on scripts/vcf_site_level_filtering_wdl/JointVcfFiltering.wdl:
```diff
@@ -237,21 +243,20 @@ task ScoreVariantAnnotations {
 			-V ~{vcf} \
 			-O ~{basename}.~{mode} \
-			--python-script ~{scoring_python_script} \
+			--model-backend ~{model_backend} \
```
Yes, I just went ahead and made the model backend an explicit parameter now. Seems like the amount of argument validation would be the same (e.g., if we implicitly toggled on a custom backend using the presence of a script argument, we'd want to check that a backend argument wasn't specified) and it might make WDLing a bit simpler?

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotations.java:
```diff
@@ -0,0 +1,361 @@
+ *     This tool is intended to be used as the first step in a variant-filtering workflow that supersedes the
+ *     {@link VariantRecalibrator} workflow. This tool extracts site-level annotations, labels, and other relevant metadata
+ *     from variant sites (or alleles, in allele-specific mode) that are or are not present in specified labeled
+ *     resource VCFs (e.g., training or calibration VCFs). The former, present sites are considered labeled; each site
```
Tweaked it a bit differently, but I agree that the original wording was awkward!

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotations.java:
```diff
@@ -0,0 +1,361 @@
+            fullName = MAXIMUM_NUMBER_OF_UNLABELED_VARIANTS_LONG_NAME,
+            doc = "Maximum number of unlabeled variants to extract. " +
+                    "If greater than zero, reservoir sampling will be used to randomly sample this number " +
+                    "of sites from input sites that are not present in the specified resources.",
```
Good idea. I think suggesting that this number be comparable to the size of the labeled training set might provide some guidance (although we won't necessarily know that number before running the tool); memory/resource considerations during training might enter into the decision as well.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotations.java:
```diff
@@ -0,0 +1,361 @@
+        rng = RandomGeneratorFactory.createRandomGenerator(new Random(reservoirSamplingRandomSeed));
+        unlabeledDataReservoir = maximumNumberOfUnlabeledVariants == 0
+                ? null
+                : new LabeledVariantAnnotationsData(annotationNames, resourceLabels, useASAnnotations, maximumNumberOfUnlabeledVariants);
```
Hmm, I suppose we could do that, but one argument for leaving things this way is that both labeled and unlabeled HDF5 files will have the same directory structure (the labels/* arrays in the unlabeled file will just be all-zero arrays), which could be nice for downstream (e.g., plotting) code.

meganshand replied:
Ah ok, that makes sense. I'd just add a comment here since I found it confusing.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotations.java:
```diff
@@ -0,0 +1,361 @@
+    private void writeUnlabeledAnnotationsToHDF5() {
+        final File outputUnlabeledAnnotationsFile = new File(outputPrefix + UNLABELED_TAG + ANNOTATIONS_HDF5_SUFFIX);
+        if (unlabeledDataReservoir.size() == 0) {
+            throw new GATKException("No unlabeled variants were present in the input VCF.");
```
Remember that this is the extraction step, so we're probably not sharding it like the scoring step.

In the event that we want to use a positive-negative approach, but happen to have no input variants that are not in the resources (which would be a pretty unusual scenario), I think we want to fail early---unless our call of the training step intelligently switches back to a positive-only approach by omitting the input of the unlabeled annotations, which would probably require additional WDL plumbing (also note that there are other arguments in the training tool, such as the calibration-sensitivity threshold, that are only appropriate for positive-negative). In any case, I'd guess this is probably a relatively unlikely scenario. What do you think?

And good call on expanding the error message---done!

meganshand replied:
Yeah that makes sense. The failure I had seen was in the scoring, not the extraction as you said. Failing early sounds good!

samuelklee replied:
Sounds good! That said, we can revisit once we add all the positive-negative stuff to the WDL and expand the tests---could uncover some edges cases at the WDL level.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/LabeledVariantAnnotationsWalker.java:
```diff
@@ -0,0 +1,382 @@
+ *   - a VCF file, containing the input variants, with labels and scores appended for those passing variant-type checks TODO + calibration-sensitivity scores + filters applied?
+ */
+@CommandLineProgramProperties(
+        // TODO
```
Oops, thanks for catching this! Hmm, actually, should I just remove `@CommandLineProgramProperties` and `@DocumentedFeature`, given that this isn't intended to be used as a standalone tool?

meganshand replied:
Yeah I think removing it makes sense.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/LabeledVariantAnnotationsWalker.java:
```diff
@@ -0,0 +1,382 @@
+            resourceLabels.addAll(trackResourceLabels);
+            logger.info( String.format("Found %s track: labels = %s", resource.getName(), trackResourceLabels));
+        }
+        resourceLabels.forEach(String::intern);
```
Hmm, I actually had a TODO here that I deleted just before opening the PR about checking whether this was necessary. Now I can't remember why I added it. Probably thought it might have an effect on the memory needed for per-site labels. Let me come back to this after doing some profiling!

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/LabeledVariantAnnotationsWalker.java:
```diff
@@ -0,0 +1,382 @@
+        return null;
+    }
+
+    // TODO maybe clean up all this Triple and metadata business with a class?
```
Yes, I'll try to break out remaining TODOs into issues.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java:
```diff
@@ -0,0 +1,625 @@
+                                                                  final VariantAnnotationsScorer variantTypeScorer,
+                                                                  final List<Double> allScores) {
+        final File variantTypeAnnotationsFile = LabeledVariantAnnotationsData.subsetAnnotationsToTemporaryFile(annotationNames, allAnnotations, isVariantType);
+        final File variantTypeScoresFile = IOUtils.createTempFile("temp", ".scores.hdf5");
```
Yes, the modeling/scorer interfaces are both file interfaces. It's just the simplest way to get things done in Python. Note that the Java BGMM will also communicate through files when it goes in (even though we could just use Java arrays for input annotations and output scores), just for the sake of keeping the tool-backend interfaces identical.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java:
```diff
@@ -0,0 +1,625 @@
+        final List<Double> scores = useASAnnotations
+                ? altAlleles.stream().map(a -> scoresIterator.next()).collect(Collectors.toList())
+                : Collections.singletonList(scoresIterator.next());
+        final double score = Collections.max(scores);
```
Yeah, this is just borrowed from ApplyVQSR. It would be nice to retain allele-level scores, but I'm not sure users would make use of them downstream. Things might also get a bit messy for non-multiallelic sites if we have annotations for both site-level and allele-level scores in the mix. I'll add an issue.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/TrainVariantAnnotationsModel.java:
```diff
@@ -0,0 +1,571 @@
+ *     This tool is intended to be used as the second step in a variant-filtering workflow that supersedes the
+ *     {@link VariantRecalibrator} workflow. Given training (and optionally, calibration) sets of site-level annotations
+ *     produced by {@link ExtractVariantAnnotations}, this tool can be used to train a model for scoring variant
+ *     calls. The outputs of the tool are TODO
```
Yup, working on it!

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/TrainVariantAnnotationsModel.java:
```diff
@@ -0,0 +1,571 @@
+        if (numTrainingAndVariantType > 0) {
+            logger.info(String.format("Training %s model with %d training sites x %d annotations %s...",
+                    variantTypeString, numTrainingAndVariantType, annotationNames.size(), annotationNames));
+            final File labeledTrainingAndVariantTypeAnnotationsFile = LabeledVariantAnnotationsData.subsetAnnotationsToTemporaryFile(annotationNames, annotations, isTrainingAndVariantType);
```
Yes, this is just a consequence of the file interface and needing to subset by variant type.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotations.java:
```diff
@@ -0,0 +1,361 @@
+ *     {@link VariantRecalibrator} workflow. This tool extracts site-level annotations, labels, and other relevant metadata
+ *     from variant sites (or alleles, in allele-specific mode) that are or are not present in specified labeled
+ *     resource VCFs (e.g., training or calibration VCFs). The former, present sites are considered labeled; each site
+ *     can have multiple labels. The latter sites are considered unlabeled and can be randomly downsampled using
```
Included this line in the above tweak---note that the original "latter" was referring to the sites that "are not present in resource VCFs," rather than the calibration sites. Confusing, indeed!

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java:
```diff
@@ -0,0 +1,625 @@
+    @Override
+    VCFHeader constructVCFHeader(final List<String> sortedLabels) {
+        final VCFHeader inputHeader = getHeaderForVariants();
+        final Set<VCFHeaderLine> inputHeaders = inputHeader.getMetaDataInSortedOrder();
```
Thanks, changed to `sortedInputHeaderMetaData`. I think this was cribbed from ApplyVQSR and I see that it also appears in other tools...

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/modeling/BGMMVariantAnnotationsModel.java:
```diff
@@ -0,0 +1,31 @@
+public final class BGMMVariantAnnotationsModel implements VariantAnnotationsModel {
+
+    public BGMMVariantAnnotationsModel(final File hyperparametersJSONFile) {
+        throw new NotImplementedException("BGMM module implemented in separate PR.");
```
Good call, changed it everywhere.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/modeling/PythonSklearnVariantAnnotationsModel.java:
```diff
@@ -0,0 +1,68 @@
+ * The lambda should check the test annotation names against the training annotation names and
+ * then return a numpy array of float-valued scores with length given by the number of data points.
+ *
+ * See org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/isolation-forest.py for an example implementation.
```
I'm OK with leaving it in the `vqsr` package for now---all of the CNN stuff is still in there, too. Maybe we can move it to its own package when we delete VQSR :)

As for whether it belongs in `src/main/python`, the answer is no with the way we currently do things in gCNV and the CNN.  Python scripts invoked by each Java tool go in `src/main/resources`, while Python packages and code called by those scripts go in `src/main/python`.

meganshand replied:
ah ha, TIL!

samuelklee commented on src/main/resources/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/isolation-forest.py:
```diff
@@ -0,0 +1,138 @@
+    print('Hyperparameters:', hyperparameters_kwargs)
+
+    print('Imputing annotations...')
+    imputer = sklearn.impute.SimpleImputer(strategy='median')
```
Hmm, interesting thought. I think at that point one would probably use a custom `PYTHON_SCRIPT` and they could have Python-side code to parse the hyperparameters JSON if needed. I will be sure to mention in the docs for the train tools that `PYTHON_IFOREST` uses median imputation, though.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ExtractVariantAnnotationsIntegrationTest.java:
```diff
@@ -0,0 +1,253 @@
+            "DP", "FS", "MQ", "MQRankSum", "QD", "ReadPosRankSum", "SOR");
+
+    private static final List<String> ALLELE_SPECIFIC_ANNOTATIONS = Arrays.asList(
+            "DP", "AS_FS", "AS_MQ", "AS_MQRankSum", "AS_QD", "AS_ReadPosRankSum", "AS_SOR");
```
Oops, good catch! Although I guess in principle one could mix non-AS/AS specific annotations if the semantics were appropriate. I'll regenerate expected results after I also address the `PG` line comment below.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/TrainVariantAnnotationsModelIntegrationTest.java:
```diff
@@ -0,0 +1,432 @@
+                    extractTag + ExtractVariantAnnotations.UNLABELED_TAG + LabeledVariantAnnotationsWalker.ANNOTATIONS_HDF5_SUFFIX);
+            final Function<ArgumentsBuilder, ArgumentsBuilder> addUnlabeledAnnotations = ab ->
+                    ADD_UNLABELED_ANNOTATIONS_HDF5.apply(ab, unlabeledAnnotationsHDF5);
+            final double calibrationSensitivityThreshold = CALIBRATION_SENSITIVITY_THRESHOLD;
```
Inlined everywhere. I guess we currently don't even need the lambda, since we use the same constant everywhere, but I left it in.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotationsIntegrationTest.java:
```diff
@@ -0,0 +1,260 @@
+                .apply(argsBuilder);
+        runCommandLine(argsBuilder);
+        Assert.assertFalse(new File(outputPrefix + ScoreVariantAnnotations.ANNOTATIONS_HDF5_SUFFIX).exists());
+        Assert.assertFalse(new File(outputPrefix + ScoreVariantAnnotations.SCORES_HDF5_SUFFIX).exists());
```
Yes, unfortunately this might be a limitation of our HDF5 wrapper (https://github.com/broadinstitute/hdf5-java-bindings/blob/master/src/main/java/org/broadinstitute/hdf5/HDF5File.java) and perhaps even the Java HDF5 library itself. Our wrapper, at least, disallows writing of annotation matrices with no elements (ideally, we'd want to write a 0 x M matrix, where M is the number of features, so that we could at least encode M). I'm not sure if this is also true at the library level (or true for double arrays at either the wrapper or library level), but we could check.

In general, some Googling shows that the treatment of HDF5 null/empty/etc. arrays has evolved a bit over time and may differ across different versions/implementations (e.g., h5py). I think we'd want to match h5py/numpy behavior, if possible, since I would guess this is what most downstream code would use.

I'm not sure if `touch` will get around this without special treatment in downstream code, since an empty text file cannot really be treated the same as an HDF5 file with empty matrices/arrays. Perhaps optional outputs would be a better option (no pun intended)?

In any case, I'll file an issue and leave this to future work. Changing our HDF5 wrapper at all seems to add a lot of overhead...it hasn't really been updated in quite some time.

meganshand replied:
Ah right, I'm not actually using the HDF5's downstream in the WDL so the empty file output didn't matter. I can update it to an optional output.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/TrainVariantAnnotationsModelIntegrationTest.java:
```diff
@@ -0,0 +1,432 @@
+                outputPrefixSNPPlusIndel + ".snp" + TrainVariantAnnotationsModel.CALIBRATION_SCORES_HDF5_SUFFIX));
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class, groups = {"python"}) // python environment is required to run tool
```
Some of these tests hit the Python package checks before the IAEs, but I removed the tag from the others---good catch.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/data/VariantType.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Logic for determining variant types was retained from VQSR.
+ */
+public enum VariantType {
```
Yes, I decided to borrow all this from VQSR for the time being to make variant classifications in evaluation bakeoffs line up exactly---no reason we have to retain it going forward. I'll file an issue to revisit.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/data/VariantType.java:
```diff
@@ -0,0 +1,49 @@
+    public static VariantType getVariantType(final VariantContext vc,
+                                             final Allele allele) {
+        if (vc.getReference().length() == allele.length()) {
+            //note that spanning deletions are considered SNPs by this logic
```
I agree. See above comment, we can revisit.

samuelklee replied:
Although also note that there's an external spanning deletion check in calling code---this was how things were also done in VQSR.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/data/VariantType.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static VariantType getVariantType(final VariantContext vc) {
+        if (vc.isSNP() || vc.isMNP()) {
```
Yeah, seems like we could use a similar switch statement. Also not sure why e.g. structural indels are treated differently. But I'll just mention this in the issue and leave as is for now, just in case there is any weird edge-case behavior that we'd change if we changed this code.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/modeling/PythonSklearnVariantAnnotationsModel.java:
```diff
@@ -0,0 +1,68 @@
+                                                       final String outputPrefix) {
+        try {
+            return new ArrayList<>(Arrays.asList(
+                    "--annotations_file=" + annotationsFile.getCanonicalPath(),
```
Done here and in the scoring class.

samuelklee commented:
OK, thanks for the thorough reviews, @meganshand and @davidbenjamin! I think I've addressed everything or left them as TODOs; I'll break these out into issues (or at least add them to the meta issue) later today. I also added the docs for the training tool.

Apologies for the slight delay, I had to get my eyes dilated on Friday morning and was completely useless for the rest of the day!

meganshand reviewed (approved):
👍  Thanks @samuelklee! The added documentation for the training tool looks great.
